File: java/54_Spiral Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        if(matrix.length == 0) {
        	return new ArrayList<Integer>();
        }
        int len = matrix.length * matrix[0].length;
        List<Integer> res = new ArrayList<Integer>();
        
        int x = 0, y = 0;
        int xLBound = 0, yLBound = 0;
        int xHBound = matrix.length, yHBound = matrix[0].length;
        
        int move = 1;
        while(res.size() < len - 1) {
        	for (; (move < 0 && y > yLBound) || (move > 0 && y < yHBound - 1); y += move) {
                res.add(matrix[x][y]);
            }
            
            for (; (move < 0 && x > xLBound) || (move > 0 && x < xHBound - 1); x += move) {
                res.add(matrix[x][y]);
            }

            if (move > 0) {
                yHBound--;
                xLBound++;
            } else {
                yLBound++;
                xHBound--;
            }
            move *= -1;
        }
        res.add(matrix[x][y]);
        return res;
    }
}

File: c++/54_SpiralMatrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        int n = matrix.size();
        if (!n) return res;
        int m = matrix[0].size();
        vector<vector<bool>> vis(n, vector<bool>(m));
        
        int dx[] ={0, 1, 0, -1}, dy[] = {1, 0, -1, 0};
        
        for (int i = 0, x = 0, y = 0, d = 0; i < n * m; i++) {
            res.push_back(matrix[x][y]);
            vis[x][y] = true;
            
            int a = x + dx[d], b = y + dy[d];
            if (a < 0 || a >= n || b < 0 || b >= m || vis[a][b]) {
                d = (d + 1) % 4;
                a = x + dx[d], b = y + dy[d];
            }
            x = a, y = b;
        }
        
        return res;
    }
};